core/src/main/scala/kafka/tools/DumpLogSegments.scala

```diff
@@ -150,6 +173,76 @@ object DumpLogSegments {
     }
   }
 
+  private def dumpTimeIndex(file: File,
+                            indexSanityOnly: Boolean,
+                            verifyOnly: Boolean,
+                            timeIndexDumpErrors: TimeIndexDumpErrors,
+                            maxMessageSize: Int) {
+    val startOffset = file.getName().split("\\.")(0).toLong
+    val logFile = new File(file.getAbsoluteFile.getParent, file.getName.split("\\.")(0) + Log.LogFileSuffix)
+    val messageSet = new FileMessageSet(logFile, false)
+    val indexFile = new File(file.getAbsoluteFile.getParent, file.getName.split("\\.")(0) + Log.IndexFileSuffix)
+    val index = new OffsetIndex(indexFile, baseOffset = startOffset)
+    val timeIndex = new TimeIndex(file, baseOffset = startOffset)
+
+    //Check that index passes sanityCheck, this is the check that determines if indexes will be rebuilt on startup or not.
+    if (indexSanityOnly) {
+      timeIndex.sanityCheck
+      println(s"$file passed sanity check.")
+      return
+    }
+
+    var prevTimestamp = Message.NoTimestamp
+    for(i <- 0 until timeIndex.entries) {
+      val entry = timeIndex.entry(i)
+      val position = index.lookup(entry.offset).position
+      var partialFileMessageSet: FileMessageSet = messageSet.read(position, Int.MaxValue)
+      val shallowIter = partialFileMessageSet.iterator
+      var maxTimestamp = Message.NoTimestamp
+      var checkedOffset = -1L
+      // We stop scan when one of the following is not met:
+      // 1. All the messages are checked
+      // 2. The max timestamp checked is greater or equals to the indexed timestamp
+      // 3. When the offset checked is greater than the next indexed offset.
+      var wrapperMessageOpt = if (shallowIter.hasNext) Some(shallowIter.next) else None
+      do {
+        wrapperMessageOpt match {
+          case Some(wrapperMessage) =>
+            val deepIter = getIterator(wrapperMessage, isDeepIteration = true)
+            for (messageAndOffset <- deepIter)
+              maxTimestamp = math.max(maxTimestamp, messageAndOffset.message.timestamp)
+            checkedOffset = wrapperMessage.offset
+            wrapperMessageOpt = if (shallowIter.hasNext) Some(shallowIter.next) else None
+          case _ =>
+        }
+      } while (wrapperMessageOpt.isDefined && maxTimestamp < entry.timestamp &&
+           (i == timeIndex.entries - 1 || timeIndex.entry(i + 1).offset > checkedOffset))
+
+      if (maxTimestamp != entry.timestamp) {
```

Comment:
Ran into an issue while testing DumpLogSegments (details can be found in the jira since PR comment doesn't support attachment). It seems that if the producer sends a batch of non-compressed messages, in LogSegment.append(), we always pass in the offset of the first message, but the offset of the largest message may not be the first. This seems to be what's failing the check in the tool. Not sure if this is the behavior that we want. If yes, we will need to document this properly, i.e., the index may point to an offset before the message that has the corresponding timestamp. But then, we have to think through other implications such as how this affects truncation.